File: packages/shared/src/positions.ts
import type { Position } from "./types/player.js";

/** Standard outfield/GK codes used on the pitch. */
export const PLAYABLE_POSITIONS = new Set<string>([
  "GK",
  "RB",
  "RWB",
  "CB",
  "LB",
  "LWB",
  "CDM",
  "CM",
  "CAM",
  "RM",
  "LM",
  "RW",
  "LW",
  "CF",
  "ST",
]);

/** Bench / squad-role codes from SoFIFA — never shown or used for slot eligibility. */
const NON_PLAYABLE = new Set(["SUB", "RES"]);

/** Map SoFIFA variant codes to the slot labels we use on the pitch. */
const POSITION_ALIASES: Record<string, Position> = {
  LCB: "CB",
  RCB: "CB",
  LCM: "CM",
  RCM: "CM",
  LDM: "CDM",
  RDM: "CDM",
  LAM: "CAM",
  RAM: "CAM",
  LS: "ST",
  RS: "ST",
  LF: "LW",
  RF: "RW",
};

/** Parse + clean a raw SoFIFA `positions` string into playable position codes. */
export function normalizePlayerPositions(raw: readonly string[] | string): Position[] {
  const parts =
    typeof raw === "string"
      ? raw.replace(/,/g, " ").split(/\s+/).filter(Boolean)
      : raw;

  const seen = new Set<Position>();
  const out: Position[] = [];

  for (const part of parts) {
    const code = part.trim().toUpperCase();
    if (!code || NON_PLAYABLE.has(code)) continue;

    const mapped = (POSITION_ALIASES[code] ?? code) as Position;
    if (!PLAYABLE_POSITIONS.has(mapped) || seen.has(mapped)) continue;
    seen.add(mapped);
    out.push(mapped);
  }

  return out;
}

File: apps/server/src/draftoff/positions.py
"""Playable position codes parsed from SoFIFA — mirrors packages/shared/src/positions.ts."""

from __future__ import annotations

PLAYABLE_POSITIONS = {
    "GK", "RB", "RWB", "CB", "LB", "LWB", "CDM", "CM", "CAM", "RM", "LM", "RW", "LW", "CF", "ST",
}

NON_PLAYABLE = {"SUB", "RES"}

POSITION_ALIASES: dict[str, str] = {
    "LCB": "CB",
    "RCB": "CB",
    "LCM": "CM",
    "RCM": "CM",
    "LDM": "CDM",
    "RDM": "CDM",
    "LAM": "CAM",
    "RAM": "CAM",
    "LS": "ST",
    "RS": "ST",
    "LF": "LW",
    "RF": "RW",
}


def normalize_player_positions(raw: list[str] | str) -> list[str]:
    if isinstance(raw, str):
        parts = raw.replace(",", " ").split()
    else:
        parts = raw

    seen: set[str] = set()
    out: list[str] = []
    for part in parts:
        code = (part or "").strip().upper()
        if not code or code in NON_PLAYABLE:
            continue
        mapped = POSITION_ALIASES.get(code, code)
        if mapped not in PLAYABLE_POSITIONS or mapped in seen:
            continue
        seen.add(mapped)
        out.append(mapped)
    return out
